### ISSUE TITLE
Allow custom querystring parsing and serialization.

### ISSUE THREAD

**Issue by threehams** [typehero/type-route#23]
`type-route` uses `querystringify` internally. This is fine when starting up a new project, but for projects stuck with existing URLs, it can cause information loss in the URL and prevent type-route from being used at all.

Example querystring:
`?thing[]=stuff1&thing[]=stuff2`
parses as
`{ "thing[]": "stuff1 }`

Can you think of a way to expand router or route definition to customize querystring parsing? I experimented a bit, but it ends up being yet another overload to `defineRoute`. For that cost, it seems like there'd need to be more benefit, like general transformation of query/params (invert control a bit per-route).

**Comment by threehams** [comment]
I've been experimenting with something that could cover this and some other features at the same time. I'm hitting some TS inference limitations, but I think I can work around them.

I haven't figured out the runtime part yet - for now I'm just seeing if TS can handle it.

```
// imagine versions for optional vs. required strings
type QueryParamAdvanced<TParsed> = {
  type: "query.param.string";
  parse: (value: string) => TParsed;
  // supply both the value and the entire query string depending on needs
  stringify: (values: { value: TParsed, queryString: string }) => string;
};

const { routes } = createRouter({
  books: defineRoute(
    {
      userId: "path.param.number",
      relatedUserIds: {
        type: "query.param.string.optional",
        stringify: (value: number[]) => value.join(","),
        parse: ({ value }) => value.split(",").map(Number)
      },
      bookIds: {
        type: "query.param.string.optional",
        stringify: (value: string[]) => qs.stringify(value),
        parse: ({ queryString }) => qs.parse(queryString).bookIds
      }
    },
    params => `/books/${params.userId}`
  )
});
```

The hopeful goal is to keep all the URL stuff hidden away, without complicating the API for simpler routes:

```
routes.books.href({
  userId: 1,
  relatedUserIds: [2, 3],
  bookIds: ["5", "6"],
});
```

**Comment by bradenhs** [comment]
I like the direction this is going and appreciate the use case you've raised. I've been thinking for a while about a way to give more flexibility to query and path parameters so they can be more than just numbers and strings. I haven't however thought about needing to support parsing things such as `?thing[]=stuff1&thing[]=stuff2` into something like `{think:['stuff1','stuff2']}`. I'd be open to the idea of a user providing their own query string serializer/deserializer at an entire router level to ensure keys are properly matched up with values and then having the ability for individual parameters to have custom serializers/deserializers for the actual value. Need to think on it some more but thanks for opening this!

**Comment by bradenhs** [comment]
This PR adds the ability to provide a custom query string serializer: https://github.com/type-route/type-route/pull/35. Any feedback on the new api would be appreciated!

**Comment by bradenhs** [comment]
Released support for this in `0.4.0`! Feel free to reopen if you feel that release doesn't address this fully.
